Add QObj class

diego-plan9 opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add it in the CHANGELOG file under Unreleased section.
⚠️ If your pull request fixes an open issue, please link to the issue.

✅ I have added the tests to cover my changes.
✅ I have updated the documentation accordingly.
✅ I have read the CONTRIBUTING document.
-->

### Summary

Add a `QObj` class, that provides a more controlled structure to the data structure that is passed between the different components instead of a plain dict. Please note this PR is not about introducing QObj fully - is the first step (the structure will remain the same) before tackling the adjustment to the specs, as mentioned in #541.

### Details and comments

Sub-tasks:
- [x] Add QObj class from previous iterations
- [x] Polish and update the class
- [x] Replace usages of the dict with QObj class
- [ ] ~Add support for versioning (and validation?) - depends on the scope of the PR~

delapuente commented on qiskit/qobj/_qobj.py:
```diff
@@ -0,0 +1,115 @@
+from ._utils import QObjType
+
+
+class QObjStructure(SimpleNamespace):
```
To be honest, I don't like the name but I'm not able to provide anything better. What about `QObjEntry` or `QObjItem`? Non-blocking, anyway.

diego-plan9 replied:
Updated!

diego-plan9 commented on qiskit/qobj/_qobj.py:
```diff
@@ -0,0 +1,115 @@
+    """
+    General QObj structure.
+    """
+    REQUIRED_ARGS = ()
```
There are indeed several options for the next steps in this regard and some related issues (how to ideally build the class from the schema, etc) that need discussion; and I would indeed like to reduce the amount of repetition in further updates.

In this case, I went with a conservative approach giving priority to the goal of having the class working and keeping the `REQUIRED_ARGS` explicit for clarity, and be able split the remaining work (for example, that change would be working on the internals) afterwards to avoid blocking - would that make sense?

diego-plan9 commented:
Still WIP and the online tests are expected to fail, but the class is now in use in the codebase. Further tuning upcoming!

diego-plan9 commented:
I'm taking the PR out of WIP as the implementation now contains both the classes and the necessary adjustments to the codebase in order to use them (ie. a `QObj` is returned by `compile()`, the local backends accept `QObj` objects and the tests operate on them as well).

Please note that the main goal of this PR was to get the classes in place in order to iterate from there, and there are still aspects of the implementation that need to be polished/decided for the next steps - some information is in the commit messages and the `TODO`s in the code itself, but basically, one (or several) follow-up PRs should take care of:
- [ ] reduce the amount of verbosity when creating subclasses (long term, it should be possible to recreate them from the .json specs themselves), ideally preserving facilities for static analysis and clarity for the user
- [ ] add validation, as it was not possible with the current data structure, and proper versioning
- [ ] update the tests so the `QObj`s built into them are not created from a dictionary
- [ ] revise the `from_dict` and `as_dict` methods, as they are currently mostly a convenience for the transition
- [ ] add more tests about the `QObj` themselves

Plus the other tasks related to the integration at #541

jaygambetta commented:
This is the correct way to start. I agree with all the things you request.  Looking forward to the day we have schema validation. :-)

ajavadia commented on qiskit/backends/ibmq/ibmqjob.py:
```diff
@@ -47,15 +47,15 @@ def __init__(self, qobj, api, is_device):
-            qobj (dict): job description
+            qobj (QObj): job description
             api (IBMQuantumExperience): IBM Q API
             is_device (bool): whether backend is a real device  # TODO: remove this after Qobj
```
Do we still need this TODO?

diego-plan9 replied:
I think so, yes - as the `Qobj` for this PR does not contain the real schema.

ajavadia commented on qiskit/qobj/_qobj.py:
```diff
@@ -0,0 +1,209 @@
+        init_args = tuple(getattr(self, key) for key in self.REQUIRED_ARGS)
+        extra_args = {key: value for key, value in self.__dict__.items()
+                      if key not in self.REQUIRED_ARGS}
+        return self.__class__, init_args, extra_args
```
There is probably a good reason, but why are you returning the required args first, then the extra args?

diego-plan9 replied:
Basically for making the serialization work as closely as possible to the `__init__` instantiation - [this method](https://docs.python.org/3/library/pickle.html#object.__reduce__) is a bit magic, but it is needed in order to have serialization working (which in turn, we need because the `Qobj`s are passed to the futures, and the default serialization fails).

ajavadia commented on qiskit/qobj/_qobj.py:
```diff
@@ -0,0 +1,209 @@
+        shots (int): number of shots.
+        backend_name (str): name of the backend.
+    """
+    REQUIRED_ARGS = ['max_credits', 'shots', 'backend_name']
```
`max_credits` should be optional, not required.
Also, `backend_name` shouldn't belong in QobjConfig at all. Basically, a Qobj should not be tied to a backend, it should be possible to, for example, compile for a device, then run the output Qobj on a simulator.

diego-plan9 replied:
Hmm, I think this comment and the rest of comments about the classes in this particular file are part of the "to be done in future PRs"! I agree with your comments - but only as soon as we have something that resembles the final qobj schemas in place. At the moment, the class hierarchy and the required attributes, etc in this class have been defined following the _current output_ of compilation on purpose, in order to start with something compatible with the rest of the codebase, and work from there.

ajavadia commented on qiskit/qobj/_qobj.py:
```diff
@@ -0,0 +1,209 @@
+
+
+# TODO: Remove when new schema is in place?
+class QObjCompiledCircuit(QObjItem):
```
This class should be called `QobjCompiledExperiment`. The idea with qobj is that it is a datastructure describing what the backend should do. And this could be either in the language of circuits/gates, or in the language of pulses. Experiment is a generic word that captures both.

I know that in our current setup we have things like `compiled_circuit` and `compiled_circuit_qasm`, and these should stay until full migration (as you have done in your `QobjExperiment` class). But we can at least define the classes now, as we would like them in the future.

ajavadia replied:
Also, I think I don't agree with your `TODO` here. This class should stay, as it is the main class describing the payload of the experiment to be executed.

ajavadia commented on qiskit/qobj/_qobj.py:
```diff
@@ -0,0 +1,209 @@
+        coupling_map (list): coupling map
+        layout (list): layout
+    """
+    REQUIRED_ARGS = ['seed', 'basis_gates', 'coupling_map', 'layout']
```
(Flagging this here, but it is fine if you want to fix in a future PR.)
`basis_gates` and `coupling_map` should not be part of a `QobjExperimentConfig`. These are properties of the backend, so the backend already knows about them and we shouldn't include them in the payload being sent to the backend. I have to think about `layout` more, but I think `layout` can also be deleted here.

ajavadia commented on qiskit/backends/local/statevector_simulator_py.py:
```diff
@@ -59,9 +60,10 @@ def _run_job(self, qobj):
-                {'name': 'snapshot', 'params': [final_state_key]})
+        for circuit in qobj.circuits:
+            circuit.compiled_circuit.operations.append(
+                QObjItem.from_dict({'name': 'snapshot', 'params': [final_state_key]})
```
I think you should create a `QobjInstruction` here rather than `QobjItem`?

diego-plan9 replied:
Nice catch! Actually needed to tweak `QobjInstruction` a bit, as the snapshot is the only instruction without `qubits`.

ajavadia commented on qiskit/backends/local/statevector_simulator_cpp.py:
```diff
@@ -45,9 +46,10 @@ def _run_job(self, qobj):
-                {'name': 'snapshot', 'params': [final_state_key]})
+        for circuit in qobj.circuits:
+            circuit.compiled_circuit.operations.append(
+                QObjItem.from_dict({'name': 'snapshot', 'params': [final_state_key]})
```
I think you should create a `QobjInstruction` here rather than `QobjItem`?

ajavadia commented on qiskit/backends/local/statevector_simulator_py.py:
```diff
@@ -87,15 +89,16 @@ def _validate(self, qobj):
-                circuit['config']['shots'] = 1
+                circuit_qobj.config.shots = 1
             for op in circuit['compiled_circuit']['operations']:
                 if op['name'] in ['measure', 'reset']:
```
Shouldn't these lines be updated also to work with object not dictionary?

diego-plan9 replied:
Yep, fixed!

ajavadia commented on qiskit/backends/local/statevector_simulator_cpp.py:
```diff
@@ -71,15 +73,16 @@ def _validate(self, qobj):
-                circuit['config']['shots'] = 1
+                circuit_qobj.config.shots = 1
             for op in circuit['compiled_circuit']['operations']:
                 if op['name'] in ['measure', 'reset']:
```
Shouldn't these lines be updated also to work with object not dictionary?

diego-plan9 commented:
Thanks @ajavadia and @ewinston ! Yes, I think as long as we are consistent, `Qobj` sounds fine - I went ahead and renamed all the classes and updated the references to the proper capitalization, and also addressed your comments, with the exceptions of the ones referring to the structure/required fields/etc of the `qiskit.qobj.*` classes. The reason is that this PR is aiming to mimic the current output of the transpiler, so those decisions are based basically on the current structure: some sort of stepping stone before fully adhering to the schema.
